src/Compilers/Core/Portable/CodeGen/LocalSlotManager.cs

```diff
@@ -246,7 +254,53 @@ private LocalDefinition DeclareLocalImpl(
         internal void FreeSlot(LocalDefinition slot)
         {
             Debug.Assert(slot.Name == null);
-            FreeSlots.Push(new LocalSignature(slot.Type, slot.Constraints), slot);
+
+            if (_nonReusableLocals?.Remove(slot) != true)
+            {
+                FreeSlots.Push(new LocalSignature(slot.Type, slot.Constraints), slot);
+            }
+        }
+
+        internal int StartScopeOfTrackingAddressedLocals()
+        {
+            Debug.Assert((_addressedLocals == null) == (_addressedLocalScopes == 0));
+            _addressedLocals ??= ArrayBuilder<LocalDefinition>.GetInstance();
+            _addressedLocalScopes++;
+            return _addressedLocals.Count;
+        }
+
+        internal void AddAddressedLocal(LocalDefinition localDef)
+        {
+            if (localDef != null)
+            {
+                _addressedLocals?.Add(localDef);
```

Comment:
>_addressedLocals?.Add(localDef); [](http://example.com/codeflow?start=16&length=32)

Should user defined locals be subject for this treatment? Presumably ref safety for them was verified during binding. Let's add a unit test with a scenario affecting a user local.

Reply:
my intuition was that this should impact temps only so share the concern about user defined locals.

Reply:
This does not have any effect on user-defined locals - they are ignored by `CodeGenerator.FreeLocal`, i.e., they are never reused anyway, regardless of whether we add them to the `_addressedLocals` and `_nonReusableLocals` lists. But I could filter them here as well, i.e., avoid adding them to those lists in the first place.

> Let's add a unit test with a scenario affecting a user local.

Not sure which exact scenario you are looking for. There is for example `RefEscapingTests.RefTemp_NoTemp` added as part of this PR.

Reply:
> But I could filter them here as well, i.e., avoid adding them to those lists in the first place if you think that would be better.

Yes, I think it would be better to be explicit about handling of user-defined locals.